<a href="https://colab.research.google.com/github/RumethR/Resume-Ner/blob/main/Cw1_w1809911_RumethRandombage.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Part A: Application Area Review

**Selected Domain: Human resources and computing.**



The domain of human resources within an organisation entails a wide variety of operations and responsibilities ranging from employee compensation to organisational structure. This poses a great opportunity for Artificial Intelligent (AI) systems to contribute towards the Human Resource Management (HRM) tasks within an organisation. One of the most common areas where AI has been highly used is within the realm of recruitment and talent acquisition. A study by the ‘Oracle’ corporation has revealed that professionals in the human resource management industry believe that AI can boost productivity among other statistics (Rahmani and Kamberaj, 2021). Other sources believe that AI will enable new application procedures for candidates, increase applicant discovery rates and even reduce vacancy times as these systems can traverse through applications faster than trained professionals (Vrontis et al., 2021). AI powered applications in HRM are also able to process large amounts of data at high speeds, faster and more accurately than any trained professional, this allows organisations to gain insights that would otherwise be impossible with manual labour. The use of these systems are also more likely to be less biassed towards factors that do not affect performance such as age, gender or race (assuming that they are developed properly). Repetitive tasks within the domain of HRM are more likely to be successfully automated with AI powered applications according to *Vrontis et al. (2021)*.

The overall effect of bias that is induced when training was studied by the influential work of *Tambe, Cappelli and Yakubovich (2019)* who observed that the data that these systems were built on had a major impact in how these systems performed. This underlying data itself has the opportunity to represent industry swings and unspoken ethics (i.e. the system might be trained on data that indicated towards hiring from a specific university), hence it is important that these systems work in tandem with trained professionals.This study also looks into how Machine Learning algorithms have fared against several different HR operations within organisations. AI powered HRM task management systems such as Quine, BenefitFocus and Jobvite are now widely available for any organisation of any scale to incorporate into their workflows (Tambe, Cappelli and Yakubovich, 2019).

On the opposite side of the spectrum, employees now have access to tools such as IBM’s ‘Blue Match’ presents viable job openings for their specific skill set, similarly AI powered tools have helped IBM to identify possible flight risks and retain employees for much longer according to their former CEO Ginni Rometty (Rosenbaum, 2019). These HRM service vendors however do not provide any information about the type of AI techniques that are used internally within their products. It is also important to note that most AI powered solutions can typically be applied on a small portion of a HRM task, for example in the task of vetting candidates these tools can only provide a suggestion or score of desirability is most cases rather than performing the entire stack of recruiting tasks that go alongside the vetting process (Ghosh, Majumder and Santosh Kumar Das, 2023). The common theme among most studies that observe the usage of AI in HRM is that they are heavily reliant on the quality and the quantity of the data that they are being built upon.

With the rise of open source Large Language Models (LLM) it is easier than ever for organisations to create a custom tailored algorithm that can perform trivial tasks within HRM such as answering frequently asked questions from employees through chatbots *Pawan Budhwar et al. (2023)* studied the use of popular generative AI tool ChatGPT in a HRM context and how it affects areas other than productivity such as employment relations, employee wellbeing and engagement. This study revealed that employees prefer tools such as ChatGPT in many scenarios that require understanding and perceiving information more than their human co-workers. However long term social, ethical and legal implications of using such technologies are difficult to quantify (Pawan Budhwar et al., 2023). Listed below are some other AI-based techniques used in the HRM context (Ghosh, Majumder and Santosh Kumar Das, 2023).

*   Turnover Prediction with Artificial Neural Networks
*   Candidate Search With Knowledge-Based Search Engines
*   Sentiment Analysis on employee engagement and relations
*   Résumé data acquisition using NLP




# Part B: Comparision and evaluation of AI Techniques used in Human Resources

As mentioned before all HRM tasks in an organization cannot be supplemented or automated by one singular system. The scope of this specific section will cover the task of utilizing Machine Learning techniques for the purpose of recruiting, vetting and staffing employees using a CV/Résumé. This project will focus on using Named Entity Recognition (NER) on CVs to identify talent that best suites a given job posting. NER is classified as a Sequence Labelling Task within the domain of Natural Language Processing (NLP), there are many different ways of addressing these sequence labelling tasks listed below are three such approaches.

### NER using Conditional Random Fields (CRF)

Using Conditional Random Fields for NER tasks is a popular machine learning based approach which uses a probabilistic model. CRFs capture dependencies among neighbouring labels in a sequence and have the ability to consider the entire input sequence when making predictions. Feature functions are used to define relationships between input features (words) and output labels (entities). These relationships have to be manually defined by the trainer. CRFs calculate the probability distribution of an output sequence (labels) given an input sequence.Transition weights in Conditional Random Fields represent the likelihood of moving from one label to another in the sequence. During the training phase, the relationships between input features, their corresponding output labels (using data from the feature functions) and the effect of the transition weights are learned from the training data.

For training a model that can perform NER using CRF text data with entities as labels are required. Necessary words in the training data sequence need to be accompanied by its corresponding entity label. Datasets for training NER models are widely available under different platforms such as Kaggle and the UCI Machine Learning Repository. One of the most common dataset and evaluation benchmark for NER is the CoNLL-2003 [dataset](https://www.clips.uantwerpen.be/conll2003/ner/). CRF based models tend to perform well as they have the ability to capture contextual information about the data. However performance is heavily dependent on the quality and relevance of engineered features, this is one of the major drawbacks of CRF based models. Feature engineering will mostly depend on the domain that the information is extracted from, and in the context of CVs people vary in the way they express information hence engineering features that capture necessary information will be quite a difficult task. Generally, CRF based models tend to be computationally efficient compared to their Deep Learning based counterparts who in turn will be better at capturing more intricate patterns within the data.

### NER using Neural Networks and pre-trained models

Currently there are many frameworks and pre-trained models that ease the task of NER. Similar to Tensorflow there are libraries that are built for performing NLP tasks. [SpaCy](https://spacy.io/usage/spacy-101) is one such Python package that offers pre-trained models for tasks such as NER. SpaCy’s default models use Convolutional Neural Networks (CNN) for the task of feature extraction and entity recognition. These models are trained on large labelled datasets hence they cover a wide range of contexts and features. By default the SpaCy package can identify entities within a  given text such as locations, dates, persons and organisations. These can be customised through a variety of API’s that are also included within the library. Similar to CRF and Transformer based models SpaCy’s CNNs can also understand contextual information regarding the input data, allowing the model to capture relationships between entities.

Efficiency of a custom built CRF based model for NER depends heavily on the training data and the quality of the engineered features. Comparatively pre-trained models from SpaCy can offer similar accuracy scores and be more computationally efficient at  the same time. However by using a pre-built model the level of customization is limited. There are other libraries that have packaged in pre-trained models for NER such as [NLTK](https://www.nltk.org/) (Natural Language Toolkit) which can offer similar functionality to SpaCy, each library has their own strengths and weaknesses.

### NER using pre-trained Transformer based models (BERT)

With the rise of pre-trained Transformer based models such as BERT and GPT, the application of these models for NLP tasks has proven to yield competitive results. The Bi-directional approach considers text preceding from both the left and the right of a given word in a sentence. This helps the model to understand a much broader context (and longer text sequences) compared to the default pre-trained models offered by SpaCy which uses a CNN. This can be especially useful in the context of parsing data from CVs which can contain ambiguous and complex sentences. There are many more pre-trained Transformer based models which can have this characteristic, these are available through the [Hugging Face](https://huggingface.co/) community platform. BERT specifically is trained on a large corpus of text using 2 unsupervised learning techniques, masked language modelling and next sentence prediction.

Unlike SpaCy, BERT is not an out-of-the-box solution for most NER tasks. This model needs to be fine-tuned in order to be used for NER purposes. This allows BERT to be domain specific and can in theory lead to better performance on very complex and domain specific NER tasks. The drawback of this level of customization is that a BERT model is more resource intensive and technically challenging to develop compared to SpaCy or NLTK. It also requires a labelled dataset for training before it can be used. Fine-tuning a BERT model requires large amounts of data to yield competitive results, which might not be publicly available in the case of CVs. In summary, BERT models can perform better than the other two approaches given that they are fine-tuned correctly, whilst SpaCy offers a better efficiency to performance ratio without requiring extensive training or customization. Although compared to training a CRF based model for NER form scratch customising a BERT model will yield better results, one drawback is it being more computationally intensive to train and use.



# Part C: Implementation

We will be using a text-based resume dataset available from Kaggle, the dataset will be accessed directly via this notebook using the Kaggle API. The credentials are hardcoded onto this notebook for ease-of-use during the viva. (Not Recommended)

In [24]:
!pip install kaggle

!mkdir ~/.kaggle
!echo '{"username":"rumethrandombage","key":"6c49cc180f0498fccf2dabb3d22f0236"}' > ~/.kaggle/kaggle.json
!chmod 600 ~/.kaggle/kaggle.json

import json

!kaggle datasets download -d dataturks/resume-entities-for-ner
!unzip resume-entities-for-ner.zip #Extracts to ./content

dataset = [] # Each element contains a json object
with open('/content/Entity Recognition in Resumes.json', 'r') as file:
    for line in file:
        dataset.append(json.loads(line))

print("Number of elements in dataset list", len(dataset))

mkdir: cannot create directory ‘/root/.kaggle’: File exists
resume-entities-for-ner.zip: Skipping, found more recently modified local copy (use --force to force download)
Archive:  resume-entities-for-ner.zip
replace Entity Recognition in Resumes.json? [y]es, [n]o, [A]ll, [N]one, [r]ename: n
Number of elements in dataset list 220


The dataset consists of 220 JSON objects. The structure of each JSON object is as follows:

> "content" -> (contains all the data in the CV as a string)

> "annotations" -> annotation label ("Name", "Skills", etc.) -> points ("start and stop index where the label is contained") - > text for label






SpaCy is packaged with a CNN based pre-built model that can perform NER for general entity types (Persons, Organizations, Locations, etc.).

In [25]:
import spacy

# Access one string of text without any annotations from the datset and pass it into the pre-trained spacy model
first_entry = dataset[0]

nlp = spacy.load("en_core_web_sm")
doc = nlp(first_entry["content"])

# Let's see how the pre-trained Model performed
for ent in doc.ents:
  print(ent.text, ent.label_)

Abhishek Jha
Application PERSON
Bengaluru GPE
Karnataka - Email PERSON
Karnataka GPE
November 2017 DATE
B.E in Information ORG
B.v.b GPE
Hubli ORG
Karnataka GPE
August 2013 to June 2017

12th DATE
Mathematics NORP
April 2011 to March 2013 DATE
10th ORDINAL
Kendriya Vidyalaya PERSON
April 2001 to March 2011 DATE
Less than 1 year DATE
Less than 1 year DATE
Database Management ORG
Less than 1 year DATE
Database Management System ORG
Less than 1 year DATE
Java PERSON
Less than 1 year DATE
• Programming PRODUCT
C++ PERSON
Java
• Oracle PERSON
• Computer Networks ORG
Linux GPE
Windows GPE
Non - Technical Skills

• Honest ORG
Calm
• Team-Player PERSON


The built-in NER model performs extremely poorly without any customisations. We can use transfer learning or we can create our own model using the NER pipeline provided by SpaCy, we will be creating a custom model for this project.

In [26]:
import spacy
from spacy.tokens import DocBin
from tqdm import tqdm # Used to show progress bars
from spacy.util import filter_spans # To make sure there are no overlapping entities


nlp = spacy.blank("en")
labels = ["Name", "College Name", "Degree", "Graduation Year", "Years of Experience", "Companies worked at", "Designation", "Skills", "Location", "Email Address"]

ner = nlp.add_pipe("ner") #only using the ner component for the pipeline provided by SpaCy

In [87]:
# import json

# Add the pre-defined labels gathered from the dataset to the pipeline
for label in labels:
  ner.add_label(label)

# Read the dataset and pre-process as needed for SpaCy
print("Example before preprocessing data")
print(dataset[0])

# Convert the data to spaCy's training data format
formatted_data = []
for entry in dataset:
    text = entry['content'].replace("\n", " ")
    entities = []
    for annot in entry['annotation']:
        if len(annot['label']) != 1 :
          continue
          print(entry)
        else:
          label = annot['label'][0]
          for point in annot['points']:
            start = point['start']
            end = point['end']

            if start == end:
              continue
            else:
              entities.append((start, end+1, label))

    if len(entities) == 0:
      continue
    else:
      formatted_data.append((text, {"entities": entities}))

print("After formatting")
print(formatted_data[0])
print(len(formatted_data))

print(formatted_data[0][1])




Example before preprocessing data
{'content': "Abhishek Jha\nApplication Development Associate - Accenture\n\nBengaluru, Karnataka - Email me on Indeed: indeed.com/r/Abhishek-Jha/10e7a8cb732bc43a\n\n• To work for an organization which provides me the opportunity to improve my skills\nand knowledge for my individual and company's growth in best possible ways.\n\nWilling to relocate to: Bangalore, Karnataka\n\nWORK EXPERIENCE\n\nApplication Development Associate\n\nAccenture -\n\nNovember 2017 to Present\n\nRole: Currently working on Chat-bot. Developing Backend Oracle PeopleSoft Queries\nfor the Bot which will be triggered based on given input. Also, Training the bot for different possible\nutterances (Both positive and negative), which will be given as\ninput by the user.\n\nEDUCATION\n\nB.E in Information science and engineering\n\nB.v.b college of engineering and technology -  Hubli, Karnataka\n\nAugust 2013 to June 2017\n\n12th in Mathematics\n\nWoodbine modern school\n\nApril 2011 

spaCy requires training data to be in 'Doc' objects. The output from the pipeline will also result in a Doc object. Finally the file will be saved in the runtime with the .spacy format.

In [88]:
# the DocBin will store the example documents
db = DocBin()
for text, annotations in tqdm(formatted_data):
    doc = nlp.make_doc(text)
    ents = []
    for start, end, label in annotations['entities']:
      span = doc.char_span(start, end, label=label)
      if span is None:
        print(start)
        print(end)
        print(label)

      ents.append(span)
    doc.ents = ents
    db.add(doc)

db.to_disk("./train.spacy")



  0%|          | 0/220 [00:00<?, ?it/s]

None
C (Less than 1 year), Database (Less than 1 year), Database Management (Less than 1 year), Database Management System (Less than 1 year), Java (Less than 1 year)
Kendriya Vidyalaya
Woodbine modern school
None
B.v.b college of engineering and technology
None
Accenture
Application Development Associate
None
Bengaluru
Accenture
Application Development Associate
Abhishek Jha


TypeError: ignored

# References

Includes references for all of the above sections of the notebook

Ghosh, S., Majumder, S. and Santosh Kumar Das. (2023). Artificial Intelligence Techniques in Human Resource Management. Apple Academic Press (CRC Press). Available from https://doi.org/10.1201/9781003328346.

<br/>

Pawan Budhwar et al. (2023). Human resource management in the age of generative artificial intelligence: Perspectives and research directions on ChatGPT. Human resource management in the age of generative artificial intelligence: Perspectives and research directions on ChatGPT, 33 (3). Available from https://doi.org/10.1111/1748-8583.12524.

<br/>

Rahmani, D. and Kamberaj, H. (2021). Implementation and Usage of Artificial Intelligence Powered Chatbots in Human Resources Management Systems. International Conference on Social and Applied Sciences. May 2021. Available from https://www.researchgate.net/profile/Hiqmet-Kamberaj/publication/351345726_Implementation_and_Usage_of_Artificial_Intelligence_Powered_Chatbots_in_Human_Resources_Management_Systems/links/60926106299bf1ad8d78e1d1/Implementation-and-Usage-of-Artificial-Intelligence-Powered-Chatbots-in-Human-Resources-Management-Systems.pdf [Accessed 19 December 2023].

<br/>

Rosenbaum, E. (2019). IBM artificial intelligence can predict with 95% accuracy which workers are about to quit their jobs. CNBC. Available from https://www.cnbc.com/2019/04/03/ibm-ai-can-predict-with-95-percent-accuracy-which-employees-will-quit.html [Accessed 1 January 2024].

<br/>

Tambe, P., Cappelli, P. and Yakubovich, V. (2019). Artificial Intelligence in Human Resources Management: Challenges and a Path Forward. California Management Review, 61 (4), 15–42. Available from https://doi.org/10.1177/0008125619867910.

<br/>

Vrontis, D. et al. (2021). Artificial intelligence, robotics, advanced technologies and human resource management: a systematic review. The International Journal of Human Resource Management, 33 (6), 1–30. Available from https://doi.org/10.1080/09585192.2020.1871398.